File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        if(!Arrays.asList(words).contains(target)){ // 만약 target이 들어있지 않다면 0 출력
            return 0;
        }
        
        boolean[] visited = new boolean[words.length];
        Queue<String> queue = new LinkedList<>();
        queue.offer(begin);
        
        while(!queue.isEmpty()){
            int size = queue.size(); // queue가 int queue이면 항목을 추가해서 횟수를 계산할수 있지만 String queue이기 때문에 따로 루프를 주어서 횟수를 계산
            for(int i = 0; i < size; i++){
                String current = queue.poll();
                
                if(current.equals(target)){ //문제 풀때 까먹었던 최종 조건 설정
                    return answer;
                }
                
                for(int j = 0; j < words.length; j++){
                    String other = words[j];
                    if(check(current, other) && !visited[j]){ //조건 검사하고 방문하지 않았다면 큐에 삽입
                        queue.offer(other);
                        visited[j] = true;
                    }
                }
            }answer++;
            
        }
        
        return answer;
    }
    private boolean check(String word, String other){ // 단어 두개를 비교해서 다른 철자가 1개인지 체크
        int count = 0;
        for(int i = 0; i < word.length(); i++ ){
            if(count >= 2){
                return false;
            }
            if(word.charAt(i) != other.charAt(i)){
                count++;
            }
        }
        if(count == 1){
            return true;
        }else{
            return false;
        }
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    if target not in words:
        return 0
    
    queue = deque([(begin, 0)])
    visited = set([begin])
    
    while queue:
        word, depth = queue.popleft()
        
        if word == target:
            return depth
        for next_word in words:
            if next_word not in visited and is_one_diff(word, next_word):
                visited.add(word)
                queue.append((next_word, depth+1))
    
    
    return 0
def is_one_diff(a, b):
    """두 단어가 정확히 한 글자만 다르면 True"""
    diff_count = sum(1 for x, y in zip(a, b) if x != y)
    return diff_count == 1